src/sync/mod.rs

```diff
@@ -139,6 +148,114 @@ pub enum BeginSyncError {
     WrongChecksum(String),
 }
 
+pub async fn maybe_end_sync(
+    store: &dag::Store,
+    maybe_end_sync_req: &MaybeEndSyncRequest,
+) -> Result<MaybeEndSyncResponse, MaybeEndSyncError> {
+    use MaybeEndSyncError::*;
+
+    // Ensure sync head is what the caller thinks it is.
+    let mut dag_write = store.write().await.map_err(WriteError)?;
+    let dag_read = dag_write.read();
+    let sync_head_hash = dag_read
+        .get_head(SYNC_HEAD_NAME)
+        .await
+        .map_err(ReadError)?
+        .ok_or(MissingSyncHead)?;
+    if sync_head_hash != maybe_end_sync_req.sync_head {
+        return Err(WrongSyncHead);
+    }
+
+    // Ensure another sync has not landed a new snapshot on the main chain.
+    let sync_snapshot = Commit::base_snapshot(&sync_head_hash, &dag_read)
+        .await
+        .map_err(NoBaseSnapshot)?;
+    let main_head_hash = dag_read
+        .get_head(db::DEFAULT_HEAD_NAME)
+        .await
+        .map_err(ReadError)?
+        .ok_or(MissingMainHead)?;
+    let main_snapshot = Commit::base_snapshot(&main_head_hash, &dag_read)
+        .await
+        .map_err(NoBaseSnapshot)?;
+    let meta = sync_snapshot.meta();
+    let sync_snapshot_basis = meta.basis_hash().ok_or(SyncSnapshotWithNoBasis)?;
+    if sync_snapshot_basis != main_snapshot.chunk().hash() {
+        return Err(OverlappingSyncs);
+    }
+
+    // Collect pending commits from the main chain and determine which
+    // of them if any need to be replayed.
+    let mut pending = Commit::pending(&main_head_hash, &dag_read)
+        .await
+        .map_err(PendingError)?;
+    let sync_head = Commit::from_hash(&sync_head_hash, &dag_read)
+        .await
+        .map_err(LoadCommitError)?;
+    pending.retain(|c| c.mutation_id() > sync_head.mutation_id());
+    // pending() gave us the pending mutations in sync-head-first order whereas
+    // caller wants them in the order to replay (lower mutation ids first).
+    pending.reverse();
+
+    // Return replay commits if any.
+    if !pending.is_empty() {
+        let mut replay_mutations: Vec<Mutation> = Vec::new();
+        for c in pending {
+            let (name, args) = match c.meta().typed() {
+                MetaTyped::Local(lm) => (
+                    lm.mutator_name().to_string(),
+                    lm.mutator_args_json().to_vec(),
+                ),
+                _ => return Err(ProgrammerError("pending mutation is not local".to_string())),
```

Comment:
Can we change the name `ProgrammerError` to `InternalError` throughout? Our customers are also programmers and might not understand the error is something we did wrong, not them.

Reply:
done in https://github.com/rocicorp/repc/pull/193